core/src/main/java/org/jruby/RubyArithmeticSequence.java

```diff
@@ -0,0 +1,565 @@
+import static org.jruby.runtime.Helpers.hashStart;
+import static org.jruby.runtime.Helpers.murmurCombine;
+import static org.jruby.runtime.Helpers.safeHash;
+
+/**
+ * Implements Enumerator::ArithmeticSequence
+ */
+@JRubyClass(name = "Enumerator::ArithmeticSequence", parent = "Enumerator")
+public class RubyArithmeticSequence extends RubyObject {
+
+    private IRubyObject begin;
+    private IRubyObject end;
+    private IRubyObject step;
+    private IRubyObject excludeEnd;
+    private IRubyObject generatedBy;
+    private String method;
+    private IRubyObject [] args;
+
+    public static RubyClass createArithmeticSequenceClass(Ruby runtime, RubyClass enumeratorModule) {
+        RubyClass sequencec = runtime.defineClassUnder("ArithmeticSequence", runtime.getObject(), new ObjectAllocator() {
+            @Override
+            public IRubyObject allocate(Ruby runtime, RubyClass klazz) {
+                return new RubyArithmeticSequence(runtime, klazz);
+            }
+        }, enumeratorModule);
+
+        sequencec.includeModule(runtime.getEnumerable());
+        sequencec.defineAnnotatedMethods(RubyArithmeticSequence.class);
+
+        RubyClass seqMetaClass = sequencec.getMetaClass();
+        seqMetaClass.undefineMethod("new");
+
+        return sequencec;
+    }
+
+    public RubyArithmeticSequence(Ruby runtime, RubyClass klass) {
+        super(runtime, klass);
+    }
+
+    public RubyArithmeticSequence(Ruby runtime, RubyClass klass, IRubyObject generatedBy, String method, IRubyObject [] args, IRubyObject begin, IRubyObject end, IRubyObject step, IRubyObject excludeEnd) {
+        super(runtime, klass);
+        this.begin = begin;
+        this.end = end;
+        this.step = step;
+        this.excludeEnd = excludeEnd;
+        this.generatedBy = generatedBy;
+        this.method = method;
+        this.args = args;
+    }
+
+    public static RubyArithmeticSequence newArithmeticSequence(ThreadContext context, IRubyObject generatedBy, String method, IRubyObject [] args, IRubyObject begin, IRubyObject end, IRubyObject step, IRubyObject excludeEnd) {
+        return new RubyArithmeticSequence(context.runtime, context.runtime.getArithmeticSequence(), generatedBy, method, args, begin, end, step, excludeEnd);
+    }
+
+    // arith_seq_each
+    @JRubyMethod
+    public IRubyObject each(ThreadContext context, Block block) {
+        IRubyObject c = begin;
+        IRubyObject e = end;
+        IRubyObject s = step;
+        IRubyObject len_1, last;
+
+        if (!block.isGiven()) {
+            return this;
+        }
+
+        if (!(step instanceof RubyComplex) && floatStep(context, c, e, s, excludeEnd.isTrue(), true, block)) {
+            return this;
+        }
+
+        if (end.isNil()) {
+            while (true) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+            //FIXME : unreachable statement
+            //return this;
```

Comment:
Maybe just return this with an inline comment stating that it is never reached? There's not really any "fix" possible here since it's an infinite loop unless terminated by an exception.